spiketoolkit/curation/threshold_silhouette_score.py

```diff
@@ -0,0 +1,276 @@
+
+        ThresholdCurator.__init__(
+            self,
+            sorting=sorting,
+            metrics_epoch=silhouette_epoch
+        )
+
+        self.threshold_sorting(
+                threshold=threshold,
+                threshold_sign=threshold_sign
+        )
+
+
+def threshold_silhouette_score(
+    sorting,
+    recording,
+    threshold,
+    threshold_sign,
+    max_spikes_for_silhouette=cgps["max_spikes_for_silhouette"]["default"],
+    n_comp=cgps["n_comp"]["default"],
+    ms_before=cgps["ms_before"]["default"],
+    ms_after=cgps["ms_after"]["default"],
+    dtype=cgps["dtype"]["default"],
+    max_spikes_per_unit=cgps["max_spikes_per_unit"]["default"],
+    recompute_info=cgps["recompute_info"]["default"],
+    max_spikes_for_pca=cgps["max_spikes_for_pca"]["default"],
+    apply_filter=cgps["apply_filter"]["default"],
+    freq_min=cgps["freq_min"]["default"],
+    freq_max=cgps["freq_max"]["default"],
+    save_features_props=cgps["save_features_props"]["default"],
+    save_as_property=cgps["save_as_property"]["default"],
+    metric_calculator=None,
+    seed=cgps["seed"]["default"],
+):
+    """
+    Excludes units based on sihouette score.
+
+    Parameters
+    ----------
+    sorting: SortingExtractor
+        The sorting extractor to be thresholded.
+    recording: RecordingExtractor
+        The recording extractor to compute silhouette with.
+    threshold:
+        The threshold for the given metric.
+    threshold_sign: str
+        If 'less', remove units scoring less than threshold.
+        If 'less_or_equal', remove units scoring less than or equal to threshold.
+        If 'greater', remove units scoring greater than threshold.
+        If 'greater_or_equal', remove units scoring greater than or equal to threshold.
+       max_spikes_for_silhouette: int
+        Max spikes to be used for silhouette metric
+    n_comp: int
+        n_compFeatures in template-gui format
+    ms_before: float
+        Time period in ms to cut waveforms before the spike events
+    ms_after: float
+        Time period in ms to cut waveforms after the spike events
+    dtype: dtype
+        The numpy dtype of the waveforms
+    max_spikes_per_unit: int
+        The maximum number of spikes to extract (default is np.inf)
+    recompute_info: bool
+        If True, will always re-extract waveforms.
+    max_spikes_for_pca: int
+        The maximum number of spikes to use to compute PCA (default is np.inf)
+    apply_filter: bool
+        If True, recording is bandpass-filtered.
+    freq_min: float
+        High-pass frequency for optional filter (default 300 Hz).
+    freq_max: float
+        Low-pass frequency for optional filter (default 6000 Hz).
+    save_features_props: bool
+        If True, save all features and properties in the sorting extractor.
+    save_as_property: bool
+        If True, the metric is saved as sorting property
+    seed: int
```

Comment:
You need metric calculator in your docstring. See threshold num spikes

Reply:
Done.